### PR TITLE
[WIP] XCFramework support

### PR DESCRIPTION
**Pull request by tmspzz**

### CODE REVIEW

File: Source/CarthageKit/BuildSettings.swift [hunk 1]
```diff
@@ -292,6 +304,11 @@ public struct BuildSettings {
 		return self["TARGET_BUILD_DIR"]
 	}
 
+	public var shouldBuildForDistribution: Result<Bool, CarthageError> {
```
Comment (bielikb) [inline Source/CarthageKit/BuildSettings.swift]: Module stability for Swift binaries:

In order to archive & create xcframework for Swift 5.1+Xcode 11 and above, there are two settings that need to be set:
```
BUILD_LIBRARY_FOR_DISTRIBUTION=NO
```
and
```
SKIP_INSTALL=NO
```

More context for the creation & integration of xcframework (apart from Apple docs) can be found also in here:
https://github.com/bielikb/xcframeworks

Comment (bielikb) [reply Source/CarthageKit/BuildSettings.swift]: Im happy to help with this PR, so if there's anything I could help with, please let me know ;)

Comment (bielikb) [reply Source/CarthageKit/BuildSettings.swift]: @tmspzz Let me know if you need any help ;)

Comment (tmspzz) [reply Source/CarthageKit/BuildSettings.swift]: @bielikb Is the direction here clear? First effort is described here 

> Allow packing per platform (i.e iOS + simulator). #2801

and it should be followed by 

> Allow packing all platforms (i.e iOS + TvOS + MacOS ...)

See https://github.com/Carthage/Carthage/issues/2890

Comment (bielikb) [reply Source/CarthageKit/BuildSettings.swift]: To support both packings you still need to pass `SKIP_INSTALL=NO` along with `BUILD_LIBRARY_FOR_DISTRIBUTION=NO` when archiving the frameworks. I haven't got across such setup in this PR, but if that's planned/implemented already, feel free to skip this comment ;)

Comment (tmspzz) [reply Source/CarthageKit/BuildSettings.swift]: `BUILD_LIBRARY_FOR_DISTRIBUTION=YES` should be there.

I don't know about `SKIP_INSTALL=NO`, I don't recall it being mentioned anywhere.

Comment (tmspzz) [reply Source/CarthageKit/BuildSettings.swift]: Very well then, skip install it is. If you plan to continue please do both "phases" separately.
I.e. first "per platform" then "all pack"

**Comment by hlineholm** [comment]
Will this be merged any time soon?

**Comment by tmspzz** [comment]
@ikesyo do you have any idea of what is going on here with the CI?

**Comment by ikesyo** [comment]
It's unfortunate that GitHub removed macOS 10.14 and Xcode 10.x from GitHub Actions' virtual environment: https://github.blog/changelog/2019-11-06-github-actions-macos-virtual-environment-updated-to-catalina

**Comment by jdhealy** [comment]
> It’s unfortunate…

…oof 😖

**Comment by antonino-u** [comment]
Any news on this?

**Comment by bielikb** [comment]
Hi @tmspzz is this work around the support the xcframework support going to be released in near time?
Thanks!

**Review by MarinaZvyagina — APPROVED** [review]

**Comment by bukira** [comment]
ive checked out this branch and built with make install but then

MB000200:CL-Frameworks crissmyers$ carthage build --create-xcframework combined --no-use-binaries
Please update to the latest Carthage version: 0.36.0. You currently are on 0.34.0
Unrecognized arguments: --create-xcframework

**Comment by tmspzz** [comment]
@bukira You're not using the carthage you just built. Probably you are using a version from `brew`. Run `brew unlink carthage` and make install again

**Comment by bukira** [comment]
Ok worked :-) can build xcframework how ever I cant build normal frameworks, I  get no shared schemes, so have to reinstall main Carthage to do that,

However when I try to build an app on my iPhone with these xcframework I get

No code signature found.

Any ideas? Do I need to code signature the repos before I build xcframeworks ?

**Comment by tmspzz** [comment]
> I cant build normal frameworks

I'll check that. Strange.

I'll check code signature but off the top of my head that should be left to the consumer. That is the app that you are using the frameworks in.

**Comment by bukira** [comment]
Yeah my code signing is fine in my app until I add the xcframeworks

I’m using the Xcode 12 compiler as well with ur branch build

**Comment by bukira** [comment]
The iOS app install fine on the simulator just not device, with no code signature found

**Comment by bukira** [comment]
Ok, found the issue, I am using xcframework as .binaryTargets in a swift package, followed the apple doc on making binaryTargets for XCFrameworks in a swift package

  targets: [
        // Targets are the basic building blocks of a package. A target can define a module or a test suite.
        // Targets can depend on other targets in this package, and on products in packages this package depends on.
        .binaryTarget(
                   name: "AWSCore",
                   url: "myURL/AWSCore.xcframework.zip",
                   checksum: "a734a40f5c79dc71542e9903daf61efa5d46f43d16c7fc8f0540097845834d6b"
               ),

But I have to then link the frameworks manually from derived data, doesn’t seem right

The Binarys show in my “Referenced Binaries” in my swift package I link to but just by linking that swift package they don’t get “embedded and signed”

Cant find much info on how to link binaryTargets referenced in a swift package

**Comment by bukira** [comment]
Seems to be a known issue with SPM and Xcode 12 embedding xcframeworks from binaryTargets and Xcode not signing them as it should 

Work around is to force sign them in a script phase

**Comment by gjeck** [comment]
I'm also running into problems building normal frameworks as mentioned by @bukira. Output is like:
```
*** Skipped building Alamofire due to the error:
Dependency "Alamofire" has no shared framework schemes for any of the platforms: tvOS, watchOS, iOS
```

**Comment by bukira** [comment]
> I'm also running into problems building normal frameworks as mentioned by @bukira. Output is like:
> 
> ```
> *** Skipped building Alamofire due to the error:
> Dependency "Alamofire" has no shared framework schemes for any of the platforms: tvOS, watchOS, iOS
> ```
> 
> Should we not be trying this out on Xcode 12? Perhaps I should try this on 11.7?

If you want to checksum an xcframework you need Xcode 12 compiler I believe

**Comment by tmspzz** [comment]
@gjeck @bukira it should be fixed. Please pull.

**Comment by gjeck** [comment]
> @gjeck @bukira it should be fixed. Please pull.

Can build normal frameworks now! Thanks @tmspzz 😄

**Comment by bcosescu** [comment]
@tmspzz Thanks for the hard work on getting this feature in Carthage. The scenario I've tried is an .xcframework attached to a Github release or served via an https server. Here is the output for the current Carthage release(0.36) and the version from your branch:

**Official Carthage Release 0.36**

1. Binary https source:
```
*** Downloading binary-only framework My at "file:///Users/bogdan.cosescu/work/https_server/My.json"
*** Downloading My.framework binary at "1.1.1"
Invalid archive - Found multiple frameworks with the same unarchiving destination:
* 	file:///var/folders/7b/d_bvnq5s5c14xdjd09szkdjcxdn9gf/T/TemporaryItems/(A%20Document%20Being%20Saved%20By%20carthage%207)/carthage-archive.YsmATH/My.xcframework/ios-x86_64-simulator/My.framework/
	file:///var/folders/7b/d_bvnq5s5c14xdjd09szkdjcxdn9gf/T/TemporaryItems/(A%20Document%20Being%20Saved%20By%20carthage%207)/carthage-archive.YsmATH/My.xcframework/ios-arm64/My.framework/ 
		to:
	file:///Users/bogdan.cosescu/work/samples/TestCarthage/Carthage/Build/iOS/My.framework/
```

2. Binary attached for Github Release:
```
*** Downloading MyRepo.framework binary at "Release1"
***  Skipped installing MyRepo.framework binary due to the error:
	"Invalid archive - Found multiple frameworks with the same unarchiving destination:
* 	file:///var/folders/7b/d_bvnq5s5c14xdjd09szkdjcxdn9gf/T/TemporaryItems/(A%20Document%20Being%20Saved%20By%20carthage%206)/carthage-archive.AEg7bC/My.xcframework/ios-x86_64-simulator/My.framework/
	file:///var/folders/7b/d_bvnq5s5c14xdjd09szkdjcxdn9gf/T/TemporaryItems/(A%20Document%20Being%20Saved%20By%20carthage%206)/carthage-archive.AEg7bC/My.xcframework/ios-arm64/My.framework/ 
		to:
	file:///Users/bogdan.cosescu/work/samples/TestCarthage/Carthage/Build/iOS/My.framework/"

    Falling back to building from the source
*** Skipped building MyRepo due to the error:
Dependency "MyRepo" has no shared framework schemes for any of the platforms: iOS
```

**Carthage build from this branch:**

1. Binary https source:
```
*** Downloading binary-only framework My at "file:///Users/bogdan.cosescu/work/https_server/My.json"
*** Downloading My.framework binary at "1.1.1"
file:///Users/bogdan.cosescu/work/samples/TestCarthage/Carthage/Build/ios/My.xcframework/ is not a valid bundle. Binary is not Mach-O or filetype is not supported.
```

2. Binary attached for Github Release:
```
*** Downloading MyRepo.framework binary at "Release1"
***  Skipped installing MyRepo.framework binary due to the error:
	"file:///Users/bogdan.cosescu/work/samples/TestCarthage/Carthage/Build/ios/My.xcframework/ is not a valid bundle. Binary is not Mach-O or filetype is not supported."

    Falling back to building from the source
*** Skipped building MyRepo due to the error:
Dependency "MyRepo" has no shared framework schemes for any of the platforms: iOS
```

Although failing, the xcframework is present in the Carthage/Build/ios folder.

Will these scenarios be covered by your PR?

**Comment by tmspzz** [comment]
As I mentioned a [comment](https://github.com/Carthage/Carthage/issues/3019#issuecomment-698553292):

- `.version` files don't work
- binary downloads of xcframewroks don't work

They will be covered as some point. Probably not for the first release in the wild of this PR.

**Comment by bukira** [comment]
You can use SPM for binary downloads? That’s my plan and what I’m doing, make the xcframeworks with Carthage thanks to this great work , and package up in SPM .binaryTarget with a fix for Xcode 12 which i am yet to test with signing them

**Comment by elliottwilliams** [comment]
Closing this since 0.37.0 with #3071 just shipped!

**Pull request closed without merging**